classes/Hook/HookDisplayBeforeBodyClosingTag.php

```diff
@@ -49,100 +44,236 @@ public function __construct(Ps_Googleanalytics $module, Context $context)
-                                    'currency' => $this->context->currency->iso_code,
-                                    'index' => (int) $key,
-                                    'item_brand' => $gacart['brand'],
-                                    'item_category' => $gacart['category'],
-                                    'item_variant' => $gacart['variant'],
-                                    'price' => (float) $gacart['price'],
-                                    'quantity' => (int) $gacart['quantity'],
-                                ],
-                            ],
-                        ];
-                        $gaScripts .= $this->module->getTools()->renderEvent(
-                            'remove_from_cart',
-                            $eventData
-                        );
-                    }
-                } elseif (is_array($gacart)) {
-                    $it = new RecursiveIteratorIterator(new RecursiveArrayIterator($gacart));
-                    foreach ($it as $v) {
-                        $gaScripts .= $v;
-                    }
-                } else {
-                    $gaScripts .= $gacart;
-                }
-            }
-
-            $ganalyticsDataHandler->manageData('', 'D');
-        }
+        // Log information about product listing
+        $this->saveInformationAboutListing();
+
+        // Flush events stored in data storage from previous pages
+        $this->outputStoredEvents();
+
+        // Add events
+        $this->renderProductListing();
+        $this->renderSearch();
+        $this->renderCartPage();
+        $this->renderBeginCheckout();
+        $this->renderLogin();
+        $this->renderRegistration();
+
+        // Output everything
+        return $gaTagHandler->generate($this->gaScripts);
+    }
 
+    /**
+     * This method renders tracking code for product listings, like category pages.
+     */
+    private function renderProductListing()
+    {
+        // Try to get product list variable
         $listing = $this->context->smarty->getTemplateVars('listing');
+        if (empty($listing['products'])) {
+            return;
+        }
+
+        // Prepare items to our format
         $productWrapper = new ProductWrapper($this->context);
-        $products = $productWrapper->wrapProductList(isset($listing['products']) ? $listing['products'] : []);
+        $items = $productWrapper->prepareItemListFromProductList($listing['products']);
+
+        // Prepare info about the list
+        $item_list_id = $this->context->controller->php_self;
+        $item_list_name = $listing['label'];
 
-        if ($controller_name == 'order' || $controller_name == 'orderopc') {
+        // Render the event
+        $eventData = [
+            'item_list_id' => $item_list_id,
+            'item_list_name' => $item_list_name,
+            'items' => $items,
+        ];
+        $this->gaScripts .= $this->module->getTools()->renderEvent(
```

Comment:
Since you are doing a refactoring, can I suggest we stop the "append something here ... something here ..." logic?

We could do this:
- create an array / an object `$GAScriptParts`
- in the code, we slowly fill in this object
- at the end we compute the final string

Easy implementation: just concatenate each part with an `implode()`
Better implementation: smarter logic similar to what a templating engine would do (like smarty or twig)